File: javascript/src/agent/root/root.test.ts
// Tests for the root dispatcher.
import { BaseAgent, type Event } from '@google/adk';
import { describe, expect, it } from 'vitest';

import { type Envelope, Kind, newEnvelope } from '../../ingest/envelope';
import { textEvent } from '../setup/events';
import { buildRootDispatcher } from './agentsSetup';
import { Dispatcher } from './root';

function env(kind: Kind): Envelope {
  return newEnvelope(kind, 'test', Buffer.alloc(0), new Date(1000));
}

// A code agent that emits one event, used to build a real runner without an LLM.
class TrivialAgent extends BaseAgent {
  protected override async *runAsyncImpl(): AsyncGenerator<Event, void> {
    yield textEvent(this.name, 'done');
  }
  protected override async *runLiveImpl(): AsyncGenerator<Event, void> {
    // not used
  }
}

describe('Dispatcher', () => {
  it('routes by kind', async () => {
    const d = new Dispatcher();
    const got: Kind[] = [];
    d.register(Kind.CronDaily, async (e) => {
      got.push(e.kind);
    });
    expect(d.handles(Kind.CronDaily)).toBe(true);
    await d.dispatch(env(Kind.CronDaily));
    expect(got).toEqual([Kind.CronDaily]);
  });

  it('no-ops an unhandled kind', async () => {
    const d = new Dispatcher();
    expect(d.handles(Kind.Lint)).toBe(false);
    await expect(d.dispatch(env(Kind.Lint))).resolves.toBeUndefined();
  });

  it('propagates a handler error', async () => {
    const d = new Dispatcher();
    d.register(Kind.CI, async () => {
      throw new Error('handler failed');
    });
    await expect(d.dispatch(env(Kind.CI))).rejects.toThrow('handler failed');
  });
});

describe('buildRootDispatcher', () => {
  it('registers and drives the daily and weekly summary handlers through real runners', async () => {
    const d = buildRootDispatcher({
      summaryDaily: new TrivialAgent({ name: 'daily' }),
      summaryWeekly: new TrivialAgent({ name: 'weekly' }),
    });
    expect(d.handles(Kind.CronDaily)).toBe(true);
    expect(d.handles(Kind.CronWeekly)).toBe(true);
    await expect(d.dispatch(env(Kind.CronDaily))).resolves.toBeUndefined();
    await expect(d.dispatch(env(Kind.CronWeekly))).resolves.toBeUndefined();
  });

  it('registers only the daily cron when no weekly agent is given', () => {
    const d = buildRootDispatcher({ summaryDaily: new TrivialAgent({ name: 'daily' }) });
    expect(d.handles(Kind.CronDaily)).toBe(true);
    expect(d.handles(Kind.CronWeekly)).toBe(false);
  });

  it('registers the fix handlers', async () => {
    const called: Partial<Record<Kind, boolean>> = {};
    const mark = async (e: Envelope) => {
      called[e.kind] = true;
    };
    const d = buildRootDispatcher({ lintKickoff: mark, coverageKickoff: mark, ciResume: mark });
    for (const k of [Kind.Lint, Kind.Coverage, Kind.CI]) {
      expect(d.handles(k)).toBe(true);
      await d.dispatch(env(k));
      expect(called[k]).toBe(true);
    }
  });

  it('registers no cron handlers without summary agents', () => {
    const d = buildRootDispatcher({ summaryDaily: null, summaryWeekly: null });
    expect(d.handles(Kind.CronDaily)).toBe(false);
    expect(d.handles(Kind.CronWeekly)).toBe(false);
  });
});

File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/root/RootTest.kt
package io.github.jkjamies.automationagent.agent.root

import com.google.adk.kt.agents.BaseAgent
import com.google.adk.kt.agents.InvocationContext
import com.google.adk.kt.events.Event
import io.github.jkjamies.automationagent.ingest.Envelope
import io.github.jkjamies.automationagent.ingest.Kind
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.Instant

private fun env(kind: Kind): Envelope = Envelope.new(kind, "test", ByteArray(0), Instant.ofEpochSecond(1))

/** A code agent that emits one event — a real runner without an LLM. */
private class TrivialAgent : BaseAgent(name = "trivial") {
    override fun runAsyncImpl(context: InvocationContext): Flow<Event> = flow {
        emit(Event(author = "trivial"))
    }
}

class RootTest : BehaviorSpec({
    Given("a dispatcher with a registered cron handler") {
        When("dispatching that kind") {
            Then("it routes to the handler") {
                val d = Dispatcher()
                var got: Kind? = null
                d.register(Kind.CRON_DAILY) { e -> got = e.kind }
                d.handles(Kind.CRON_DAILY) shouldBe true
                d.dispatch(env(Kind.CRON_DAILY))
                got shouldBe Kind.CRON_DAILY
            }
        }
    }

    Given("a dispatcher with no handler for a kind") {
        When("dispatching it") {
            Then("it is a silent no-op") {
                val d = Dispatcher()
                d.handles(Kind.LINT) shouldBe false
                d.dispatch(env(Kind.LINT)) // must not throw
            }
        }
    }

    Given("a handler that throws") {
        When("dispatching its kind") {
            Then("the error propagates") {
                val d = Dispatcher()
                d.register(Kind.CI) { throw RuntimeException("handler failed") }
                shouldThrow<RuntimeException> { d.dispatch(env(Kind.CI)) }
            }
        }
    }

    Given("a root dispatcher built with a summary agent") {
        When("inspecting and dispatching cron kinds") {
            Then("cron kinds route to the summary workflow") {
                val d = buildRootDispatcher(RootDeps(summaryAgent = TrivialAgent()))
                d.handles(Kind.CRON_DAILY) shouldBe true
                d.handles(Kind.CRON_WEEKLY) shouldBe true
                d.dispatch(env(Kind.CRON_DAILY)) // drives a real runner, must not throw
            }
        }
    }

    Given("a root dispatcher built with fix handlers") {
        When("dispatching lint, coverage and ci") {
            Then("each registered handler is invoked") {
                val called = mutableSetOf<Kind>()
                val mark = Handler { e -> called += e.kind }
                val d = buildRootDispatcher(RootDeps(lintKickoff = mark, coverageKickoff = mark, ciResume = mark))
                d.handles(Kind.LINT) shouldBe true
                d.handles(Kind.COVERAGE) shouldBe true
                d.handles(Kind.CI) shouldBe true
                for (k in listOf(Kind.LINT, Kind.COVERAGE, Kind.CI)) d.dispatch(env(k))
                called shouldBe setOf(Kind.LINT, Kind.COVERAGE, Kind.CI)
            }
        }
    }

    Given("a root dispatcher built without a summary agent") {
        When("checking cron kinds") {
            Then("they are unhandled") {
                val d = buildRootDispatcher(RootDeps(summaryAgent = null))
                d.handles(Kind.CRON_DAILY) shouldBe false
            }
        }
    }
})
